sdk/python/kfp/containers/_component_builder.py

```diff
@@ -248,43 +308,132 @@ def build_python_component(
+        # output.
+        program_args.append('--{}{}'.format(
+            component_input.name,
+            entrypoint.OUTPUT_NAME_SUFFIX
+        ))
+        program_args.append(_structures.InputOutputPortNamePlaceholder(
+            input_name=component_input.name))
+
+      elif component_input._passing_style is None:
+        program_args.append('--{}{}'.format(
+            component_input.name,
+            entrypoint.ARGO_PARAM_SUFFIX
+        ))
+        program_args.append(_structures.InputValuePlaceholder(
+            input_name=component_input.name))
+      else:
+        raise TypeError(
+            'Only Input/OutputArtifact and parameter annotations '
+            'are supported in V2 components. '
+            'Got %s' % component_input._passing_style)
+
+    for component_output in component_spec.outputs or []:
+      if component_output._passing_style == components.OutputArtifact:
+        # For each output artifact, there'll be possibly 3 arguments passed to
+        # the user program:
+        # 1. {name of the artifact}_output_path: The actual path, or uri, of the
+        #    input artifact.
+        program_args.append('--{}{}'.format(
+            component_output.name,
+            entrypoint.OUTPUT_ARTIFACT_PATH_SUFFIX
+        ))
+        program_args.append(
+            _structures.OutputUriPlaceholder(
+                output_name=component_output.name))
+      elif component_output._passing_style is not None:
+        raise TypeError(
+            'Only Input/OutputArtifact and parameter annotations '
+            'are supported in V2 components. '
+            'Got %s' % component_output._passing_style)
+
+    program_args.append('--pipeline_context')
+    program_args.append(dsl.RUN_ID_PLACEHOLDER)
+    program_args.append('--{}'.format(entrypoint.FN_NAME_ARG))
+    program_args.append(component_func.__name__)
+
+    component_spec.implementation.container.args = program_args
+  else:
+    program_code = _purge_program_launching_code(
+        commands=command_line_args,
+        python_version=python_version,
+        entrypoint_container_path='/' + program_path)
 
   arc_docker_filename = 'Dockerfile'
   arc_requirement_filename = 'requirements.txt'
 
   with tempfile.TemporaryDirectory() as local_build_dir:
     # Write the program code to a file in the context directory
-    local_python_filepath = os.path.join(local_build_dir, program_rel_path)
+    local_python_filepath = os.path.join(local_build_dir, program_path)
     os.makedirs(os.path.dirname(local_python_filepath), exist_ok=True)
+
     with open(local_python_filepath, 'w') as f:
       f.write(program_code)
 
     # Generate the python package requirements file in the context directory
     local_requirement_filepath = os.path.join(local_build_dir, arc_requirement_filename)
+    if is_v2:
+      # For v2 components, KFP are expected to be packed in the container.
+      dependency.append(VersionedDependency(name='kfp'))
```

Comment:
nit: do we need to specify, if that's possible, a minimum version of kfp?

Reply:
Good point. Added 1.4.0 as a lower bound for now.